### ISSUE TITLE
Function invocation broken on Windows

### ISSUE THREAD

**Issue by jferrettiboke** [graphcool/framework#1031]
**Current behavior**

I am using the `email-password` template but I get all the time `null` as response. The error message is `{"error":"Function returned invalid status code: 0. Raw body: empty.last"}`.

I am using `graphcool/0.8.0 (windows-x64) node-v8.6.0`.

**Expected behavior?**

It should create a user and return the token.

**Comment by marktani** [comment]
Thanks a lot for bringing this up! We'll roll out a fix for this later today 🙂

**Comment by cameronk** [comment]
I'm experiencing the same issue with the email-password template; same error message gets returned for the `hello` function as well.

Seems to be irrespective of what's in the function - I played with `src/hello.js` and removed the async functionality, etc.

**Comment by adamdawkins** [comment]
I'm getting the same error with the `pay` function in the `full-example`

**Comment by marktani** [comment]
Do you still get this error on `0.8.2`? What are the steps for reproduction?

**Comment by jferrettiboke** [comment]
@marktani It is not possible to get `0.8.2`. Latest version with `graphcool@next` is `0.8.1`. Currently I am have `graphcool/0.8.1 (windows-x64) node-v8.6.0`. Does this version solve the issue?

**Comment by jferrettiboke** [comment]
@marktani It is not possible to get `0.8.2`. Latest version with `graphcool@next` is `0.8.1`. Currently I have `graphcool/0.8.1 (windows-x64) node-v8.6.0`. Does this version solve the issue?

**Comment by marktani** [comment]
Sorry, it's already actually `0.8.5` 🙂 

```
npm install -g graphcool
```

**Comment by jferrettiboke** [comment]
I installed `0.8.5` and I did everything from zero and it is still returning the same error. I followed all the steps mentioned in the [`email-password` template](https://github.com/graphcool/templates/tree/master/auth/email-password).

**Comment by marktani** [comment]
What GraphQL operation triggers the above error message?

**Comment by jferrettiboke** [comment]
I get the error when I attempt to sign up a user.

```graphql
mutation {
  signupUser(email: "test@test.com", password: "123456") {
    id
    token
  }
}
```

**Comment by marktani** [comment]
I cannot reproduce this issue. Here's what I do:

* `graphcool init`
* `graphcool add-template graphcool/templates/auth/email-password
* uncomment the added `functions` in `graphcool.yml` and manually add the `email` and `password` fields in to `User` in `types.graphql`
* `npm install`
* `graphcool deploy` -> deploy to `shared-eu-west-1`
* `graphcool playground`
* run `signupUser` mutation

These are all the commands I run & I don't run any other commands. Please provide a similar step-by-step reproduction.

**Comment by jferrettiboke** [comment]
@marktani I did exactly the same than you. At the moment, I am using Windows 10. Do you think it can be part of the issue when `graphcool` prepares all the functions for uploading?

There are more people like @cameronk and @adamdawkins getting this error as well. It would be nice if they can reproduce their own ways in order to know more about this issue. I would like to know what OS they are using.

**Comment by kbrandwijk** [comment]
I can also reproduce this on Windows 10, following the exact steps from https://github.com/graphcool/framework/issues/1031#issuecomment-340209901

**Comment by kbrandwijk** [comment]
Also, this has nothing to do with using templates, node_modules or anything like that.
Here's a very minimal repro:

`gc init` 
`gc deploy` -> eu

query hello -> same error.

**Comment by kbrandwijk** [comment]
So I think this is still a duplicate of https://github.com/graphcool/framework/issues/885

**Comment by Tav0** [comment]
This issue occurs for me in MacOS.

TLDR: I deployed my functions to a new project and now the function that was giving me this error works.

When I deployed to an old project. Not sure if it is the age of project, but the reason why I believe it got something with past projects is because I deployed to a new project and everything is back to normal again.

I was getting this error with only one function at my previous old project, but now the function works in my new project. I deployed my graphcool folder straight to the cloud which also created the new project by using the `gc deploy --interactive`

Another issue I was having on the 27th was that npm couldn't find `graphcool` package while trying to update it. I fixed this error by `graphcool@latest`.

Hopefully this helps to fix this bug.

**Comment by Tav0** [comment]
After doing `gc deploy -w` now this project doesn't work either...and I tried creating and deploying a new project, but it won't work anymore. =/

**Comment by marktani** [comment]
Can others confirm that this error is somehow triggered by doing `gx deploy -w`?

**Comment by kbrandwijk** [comment]
Nope, I had this on cloud deployment without ever using `-w`.

**Comment by Tav0** [comment]
What I don't understand for me local deployment works totally fine.

**Comment by pie6k** [comment]
I had the same problem. For me reason was I was trying to perform mutation like

```
client.request(`mutation{ ... }`)

```

with incorrect mutation code (missing field etc). It was propably throwing some error that was not handled by me with `try catch` and resulted with `{"error":"Function returned invalid status code: 0. Raw body: empty.last"}`

When I've added try catch like

```
try {
  client.request(`mutation{ ... }`)
} catch(err) {
  return {error: err.message};
}
```

It started to work again.

**Comment by cameronk** [comment]
Like @kbrandwijk, I can also reproduce this on Windows 10, following the steps in #1031 (comment).

Here's the commands I entered, and their responses: https://pastebin.com/QaLK6u5g
The first deploy broke, because I forgot to define the reverse relation on Post. Here's my types.graphql:

```
type Post @model {
  id: ID! @isUnique
  title: String!

  # Every relation also required a back-relation (to determine 1:1, 1:n or n:m)
  author: User! @relation(name: "UserPosts")
}

# added by email-password template: (please uncomment)
type User @model {
  # Required system field:
  id: ID! @isUnique # read-only (managed by Graphcool)

  # Optional system fields (remove if not needed):
  createdAt: DateTime! # read-only (managed by Graphcool)
  updatedAt: DateTime! # read-only (managed by Graphcool)

  email: String! @isUnique
  password: String!


  posts: [Post!]! @relation(name: "UserPosts")
}
```

After opening the playground, I executed the following:
```
query {
  hello {
    message
  }
}
```

which returned:
```
{
  "data": {
    "hello": null
  },
  "errors": [
    {
      "locations": [
        {
          "line": 3,
          "column": 3
        }
      ],
      "path": [
        "hello"
      ],
      "code": 5000,
      "message": "A function returned an unhandled error. Please check the logs for executionId 'us-west-2:simple:cj9ellb78kfaz0128d5o42qp5'",
      "requestId": "us-west-2:simple:cj9ellb78kfaz0128d5o42qp5"
    }
  ]
}
```

In the console, I see this error:
```{"error":"Function returned invalid status code: 0. Raw body: empty.last"}```

**Comment by jurosh** [comment]
I've also tried basic steps:
```
graphcool init
npm install
graphcool deploy
```

But basic query: `{  hello(name: "test") {    message  } } `  is returning error: 

```
...
"message": "A function returned an unhandled error. Please check the logs for executionId 'eu-west-1:simple:cj9eoqi5422b90165opl1midg'",
"requestId": "eu-west-1:simple:cj9eoqi5422b90165opl1midg"
```

And in logs there is already discussed response:
```
{"error":"Function returned invalid status code: 0. Raw body: empty.last"}
```

Details: `Windows 10 using graphcool/0.8.5 (windows-x64) node-v7.6.0`

**Comment by fabien0102** [comment]
Same for me on my project -> https://github.com/fabien0102/climbing-app/blob/graphcool-framework/graphcool/src/update-route-stats/update-route-stats.ts

Try with `graphcool/0.7.0 (windows-x64) node-v8.1.3` and `graphcool/0.8.5 (windows-x64) node-v8.1.3`

I don't know if it can help, but it's more debug info :wink:

**Comment by marktani** [comment]
We believe to have found a fix for this that will be released in the next version.

I could confirm this is an issue unrelated to the OS.

**Comment by timsuchanek** [comment]
The problem was Error objects that contain circular references. A fix will be released in 0.9

**Comment by mekwall** [comment]
@timsuchanek Do you have a commit somewhere that fixes this? This is currently blocking me from continue work on auth0 integration so would like to fix it temporarily until 0.9 is released.

**Comment by Tav0** [comment]
This is seriously a big blocker for us as well. I don't understand why there wouldn't be a temp fix?

**Comment by Tav0** [comment]
Apparently for me it was because of destructuring 

Luis Villavicencio

**Comment by marktani** [comment]
I cannot reproduce this error on `graphcool/0.9.0-alpha.2 (darwin-x64) node-v8.3.0` anymore. 

However, it looks like this error message still occurs on `graphcool/0.9.0-alpha.2 (windows-x64) node-v8.1.3`, and it is *not* triggered by circular objects, but by something else.

**Comment by kbrandwijk** [comment]
I'm on Windows, so I'll try to reproduce using 0.9.0.

**Comment by kbrandwijk** [comment]
I can confirm this is still an issue on Windows. Followed the exact same steps, still getting:
```
2017-11-07T14:20:43.172Z 344ms FAILURE {
  "error": "Function returned invalid status code: 0. Raw body: empty.last"
}
```
graphcool/0.9.0-alpha.2 (windows-x64) node-v9.0.0 (same error on node 8)

**Comment by kbrandwijk** [comment]
Additional testing: just `gc init` -> `gc deploy` -> `gc playground` -> `{ hello(name: "me") { message } }` still causes the same error. Any function causes this on Windows.

**Comment by kbrandwijk** [comment]
Screenshot from Windows (first), then Linux (second):
![image](https://user-images.githubusercontent.com/852069/32499072-c9d099c0-c3d1-11e7-8603-016cddb406ec.png)

**Comment by SindreSvendby** [comment]
I had this issue now, but after reading up here and testing a bit,. I was able to fix it by removing all _console.logs_
I did not have any circular refs, but if was multiple layers deep, and the objects could look like each other so a "stupid" circular ref checker could be fooled 

So tip: 
* remove all `console.log` or other usage of `console` and see if you have the problem still

**Comment by kbrandwijk** [comment]
Unfortunately, no `console.log` statements in the minimal repro example https://github.com/graphcool/framework/issues/1031#issuecomment-342498341

**Comment by alvincrespo** [comment]
I'm also getting this error. I created a repo to showcase my issue. 

https://github.com/alvincrespo/graphcool-exp

Any help would be awesome!

**Comment by marktani** [comment]
It looks like the circular object problem was the same symptom, but a different cause.

My current understanding is that invoking any function that was deployed from a Windows machine results in an unhandled function error.

**Comment by alvincrespo** [comment]
@marktani Just a heads up that I was deploying from a Mac. Not sure what information would helpful though?

**Comment by emipc** [comment]
I also got stuck with this issue, exactly when I was following the HTG tutorial and trying to create a new user.

I get this error in the console when clicking the "create account" button:

> bundle.js:3586 Uncaught (in promise) Error: GraphQL error: A function returned an unhandled error. Please check the logs for executionId 'eu-west-1:simple:cj9rg0kgj8l3d01314ctk5gts'
>     at new ApolloError (bundle.js:3586)
>     at Object.next (bundle.js:2469)
>     at SubscriptionObserver.next (bundle.js:54122)
>     at bundle.js:4018
>     at <anonymous>

In the function log I see this error output:	

> {"error":"Function returned invalid status code: 0. Raw body: empty.last"}

BUT, if I go to the playground and write the createUser mutation and set the variables, it WORKS. I've no idea of what's the difference!

I'm using graphcool/0.8.5 (windows-x64) node-v8.1.3 on Windows 10.

**Comment by alvincrespo** [comment]
Here is what I did to reproduce:

`npm i -g graphcool`

`mkdir playground && cd playground`

`graphcool init derpin`

`cd derpin`

`graphcool local up`

```
? Please choose the cluster you want to deploy to

  Shared Clusters:
  shared-eu-west-1
  shared-ap-northeast-1
  shared-us-west-2

  Local (docker):
❯ local
```

I chose local.

```
? Please choose the target name (dev)
```

I pressed enter.

```
? Please choose the service name (derpin)
```

I pressed enter.

```
graphcool deploy
```


```
graphcool  playground
```

And then I try:

```
query {
  hello(name:"Alvin") {
    message
  }
}
```

and I get:

```

A function returned an unhandled error.
{
  "data": {
    "hello": null
  },
  "errors": [
    {
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "hello"
      ],
      "code": 5000,
      "message": "A function returned an unhandled error. Please check the logs for executionId 'local:simple:cj9rg45ha003c0125qc3rijod'",
      "requestId": "local:simple:cj9rg45ha003c0125qc3rijod"
    }
  ]
}
```

**Comment by alvincrespo** [comment]
My environment:

```
➜  derpin node-v
zsh: command not found: node-v
➜  derpin node -v
v8.7.0
➜  derpin npm -v
5.5.1
➜  derpin docker -v
Docker version 17.09.0-ce, build afdb6d4
```

I'm on macOS Sierra (10.12.6)

**Comment by fabien0102** [comment]
Just for information sharing: I have done a little test to know if the function was called or not => Definitively not!
![image](https://user-images.githubusercontent.com/1761469/32575549-a56e1aea-c4d4-11e7-8e4d-2f23df63e99d.png)

```bash
$ graphcool --version
graphcool/0.9.0 (windows-x64) node-v8.9.0
```

Note: I have test the `log` function outside of graphcool :wink: / and I tried with a simple string as `content`.

**Comment by Julianojc** [comment]
I tried to recreate this auth function <script src="https://gist.github.com/Julianojc/b1059d1727d7c96d778ef10b8c29961f.js"></script> with email directly on the remote console, when I click on "Test Run" it does not return any error, however the fuction is not saved, it is only several minutes loading .... if I use CLI Deploy mode I get the same error mentioned above code: 5000

** The code used in the tests is the same as a YouTube video posted by graphcool, where the function works normally by the console, but here it only returns an error.

**Comment by alvincrespo** [comment]
So now things seem to be working after I tried creating a new app. Instead of using `local` I used `shared-eu-west-1`. I'm brand new to this, so not sure what the differences are but this seemed to work for me.

**Comment by Toosick** [comment]
I was able to reproduce & fix the error multiple times. Don't know what is causing it. I did this with the BaaS.

**Version**
`graphcool/0.9.0 (windows-x64) node-v6.10.2`

**Reproduce (uses very basic hello example)**
`gc init`
`gc deploy`

**Fix**
* Move the hello.js into the same folder as graphcool.yml
* Change `code: src/hello.js` to `code: hello.js`
* `gc deploy`
* Works 😎

**Comment by fabien0102** [comment]
Just moving the handler works! This issue seams not impact the schema file

The working fix in my case (just for one function):
```diff
diff --git a/graphcool/graphcool.yml b/graphcool/graphcool.yml
index 4544dc0..4c68148 100644
--- a/graphcool/graphcool.yml
+++ b/graphcool/graphcool.yml
@@ -19,7 +19,7 @@ functions:
   loggedInUser:
     handler:
       code:
-        src: ./src/auth0/loggedInUser.ts
+        src: loggedInUser.ts
     type: resolver
     schema: ./src/auth0/loggedInUser.graphql
 ```

**Comment by kbrandwijk** [comment]
We knew already in #885 that it was due to file path escaping in Windows: https://github.com/graphcool/framework/issues/885#issuecomment-337814173. The base path is escaped correctly, but the subfolder isn't.
That's why I didn't understand the reasoning about Error objects and circular references to begin with.

**Comment by Julianojc** [comment]
As commented above, changing the files to the same folder as the graphcool.yml file and changing the paths in the same file ... everything works again ;)

**Comment by marktani** [comment]
This is now released in `0.9.1`: https://github.com/graphcool/framework/releases/tag/0.9.1 🙂

**Comment by jtmthf** [comment]
0.9.1 didn't fix anything for me on Windows 10 x64, but moving the resolver function to the root did fix it.

**Comment by kbrandwijk** [comment]
@jtmthf I'm also on Win 10 x64. What are the steps you followed?

**Comment by kbrandwijk** [comment]
Somehow, it seems the fix wasn't released in 0.9.1...

**Comment by fabien0102** [comment]
![graphcool-loading](https://user-images.githubusercontent.com/1761469/32682064-e01f6af0-c672-11e7-93c3-1b1ac1495b5f.gif) 20 minutes… So not really working for me 😢

**Comment by kbrandwijk** [comment]
@fabien0102 Can you retry with `set DEBUG=*`?

**Comment by dukuo** [comment]
I'm having the same issue even after upgrading to `0.9.1`. Windows 10 x64 trying to run `signUpUser` mutation

**Comment by kbrandwijk** [comment]
@dukuo As mentioned above, the fix somehow didn't get included in `0.9.1`.

**Comment by Vipsey** [comment]
Same problems here.... I'm also running the same system Win 10 x64

**Comment by Vipsey** [comment]
Is there some way to fix this quickly? Should I try to build and deploy on my mac? Im in quite some hurry with the development for a client...

**Comment by dukuo** [comment]
Oh and the fix of putting all the files at the same location is not working for me either. In my case my `User` type requires more than the 2 expected values stated in  the `email/password` template, but it should point out that and not the raw body error, right?

**Comment by Vipsey** [comment]
@dukuo, I tried the same thing with change the directories, didn't work for me neither. My signupUser, return just blank null now.

**Comment by kbrandwijk** [comment]
You can try making the changes manually in your global node_modules folder... https://github.com/graphcool/framework/pull/1227 It's just 2 words in 2 lines...

**Comment by dukuo** [comment]
@kbrandwijk I did try but I only have a `Bundle.d.ts`.. I'm sure i'm missing something obvious but couldn't find it nonetheless.

**Comment by Vipsey** [comment]
@dukuo Same here...

**Comment by kbrandwijk** [comment]
@dukuo you should also have a `Bundle.js`:
Line 66:
```diff
-this.getBuildFileName = function (src) { return path.join(src.replace(/\.ts$/, '.js')); };
+this.getBuildFileName = function (src) { return path.posix.join(src.replace(/\.ts$/, '.js')); };
```
Line 254:
```diff
-var buildFileName = path.join(_this.buildDir, _this.getBuildFileName(src));
+var buildFileName = path.posix.join(_this.buildDir, _this.getBuildFileName(src));
```

**Comment by Vipsey** [comment]
![uten navn](https://user-images.githubusercontent.com/16919580/32684009-ea8b2e18-c67f-11e7-83fc-3a3187989f04.png)

I did the changes, manually, but I'm reciving null in return from the mutation. I cant figure out why...? Am I missing some headers?

**Comment by kbrandwijk** [comment]
@Vipsey At least the deployment from Windows works if you don't get the error anymore. This is a different issue, related to your mutation, not to Windows deployment. Can you create a new post on the Forum for more specifics about the issue with your mutation?

**Comment by fabien0102** [comment]
My bad, it was just my internet before, btw always the `A function returned an unhandled error.` error in production.

My graphcool output log:
https://gist.github.com/fabien0102/44fd834dc972ff91f4fa77df55cac5bd

I think it's a path issue (https://gist.github.com/fabien0102/44fd834dc972ff91f4fa77df55cac5bd#file-output-sh-L155), maybe use [slash](https://www.npmjs.com/package/slash) to fix this?

**Comment by kbrandwijk** [comment]
@fabien0102 You're having the same issue as the rest of us. `0.9.1` didn't contain the fix. You can wait for a new release, or make the changes manually as described in https://github.com/graphcool/framework/issues/1031#issuecomment-343621637.

**Comment by timsuchanek** [comment]
@kbrandwijk @fabien0102 @jferrettiboke @Vipsey @dukuo (I hope I didn't forget anyone):
My apologies for `0.9.1` not being updated correctly. It was a packaging issue.
I just released `0.9.2` and tested it myself, the `Bundler.js` is updated correctly.
Can you please confirm that? Thanks!

**Comment by kbrandwijk** [comment]
@timsuchanek Just retested on Win 10 Pro x64. `0.9.2.` contains the fix, and it works like it should.

**Comment by timsuchanek** [comment]
@kbrandwijk thanks for testing!

**Comment by alvincrespo** [comment]
I'm still encountering a similar issue on MacOS:

![](https://ac-screenshots.s3.amazonaws.com/Playground_-_httplocalhost60000simplev1cj9vrj3hi000f0125aa10u3p4_2017-11-11_15-08-05.jpg)

Running through the same steps as https://github.com/graphcool/framework/issues/1031#issuecomment-342937551 :(

**Comment by kbrandwijk** [comment]
@alvincrespo Can you run your `graphcool deploy` with `set DEBUG=*` and share your log?

**Comment by jtmthf** [comment]
@kbrandwijk I can confirm that 0.9.2 fixes the issue for me.

**Comment by alvincrespo** [comment]
@kbrandwijk I just created a new app called `derpity` and ran it with `DEBUG=*`:

```
➜  Desktop graphcool init derpity
Creating a new Graphcool service in derpity... ✔

Written files:
├─ types.graphql
├─ src
│  ├─ hello.js
│  └─ hello.graphql
├─ graphcool.yml
└─ package.json

To get started, cd into the new directory:
  cd derpity

To deploy your Graphcool service:
  graphcool deploy

To start your local Graphcool cluster:
  graphcool local up

To add facebook authentication to your service:
  graphcool add-template auth/facebook

You can find further instructions in the graphcool.yml file,
which is the central project configuration.

➜  Desktop cd derpity
➜  derpity cd derpity
autoenv_cd:chdir:1: no such file or directory: derpity
autoenv_cd:cd:1: no such file or directory: derpity
➜  derpity graphcool loal up
 ▸    loal up is not a graphcool command.
 ▸    Perhaps you meant local up
 ▸    Run graphcool help for a list of available commands.

Get in touch if you need help: https://www.graph.cool/forum
To get more detailed output, run $ export DEBUG="*"
➜  derpity graphcool local up
Running local Graphcool cluster at http://localhost:60000
Running local FaaS runtime at http://localhost:60050
This may take several minutes
$ docker-compose up -d --remove-orphans

docker   local_localfaas_1 is up-to-date
docker   local_graphcool-db_1 is up-to-date
docker   local_graphcool_1 is up-to-date

Waiting for Graphcool to initialize. This can take several minutes... ✔

Success! Cluster local already exists and is up-to-date.
To get started, execute

  $ graphcool init
  $ graphcool deploy

➜  derpity DEBUG=* graphcool deploy
  config CWD /Users/alvincrespo/Desktop/derpity +0ms
  config HOME /Users/alvincrespo +2ms
  config definitionDir /Users/alvincrespo/Desktop/derpity +0ms
  config definitionPath /Users/alvincrespo/Desktop/derpity/graphcool.yml +0ms
  config homepath /Users/alvincrespo/.graphcoolrc +0ms
  config localRCPath /Users/alvincrespo/Desktop/derpity/.graphcoolrc +1ms
  config globalRCPath /Users/alvincrespo/.graphcoolrc +0ms
  cli command id deploy +0ms
  cli:plugincache Got plugin from cache +0ms
  cli:plugincache /Users/alvincrespo/Library/Caches/graphcool/plugins.json +0ms
  cli:plugincache Got plugin from cache +1ms
  cli:plugincache /Users/alvincrespo/Library/Caches/graphcool/plugins.json +0ms
  plugins findCommand graphcool-cli-core +0ms
  plugin requiring command +0ms
  cli-engine:plugins:manager requiring /usr/local/lib/node_modules/graphcool/node_modules/graphcool-cli-core +0ms
  cli-engine:plugins:manager required +296ms
  plugin required command +298ms
  deploy run +0ms
? Please choose the cluster you want to deploy to

local

? Please choose the target name dev
Creating service derpity locally in cluster local...  client choosing clusterEndpoint http://localhost:60000/system +0ms
  client Sending query +0ms
  client       mutation addProject($name: String!, $alias: String, $region: Region, $config: String) {
  client         addProject(input: {
  client           name: $name,
  client           alias: $alias,
  client           region: $region,
  client           clientMutationId: "static"
  client           config: $config
  client         }) {
  client           project {
  client             ...RemoteProject
  client           }
  client         }
  client       }
  client
  client   fragment RemoteProject on Project {
  client     id
  client     name
  client     schema
  client     alias
  client     region
  client     isEjected
  client     projectDefinitionWithFileContent
  client   }
  client
  client        +0ms
  client { name: 'derpity',
  client   alias: undefined,
  client   region: 'EU_WEST_1',
  client   config: '{"modules":[{"name":"","content":"types: ./types.graphql\\n","files":{"./types.graphql":""}}]}' } +1ms
 ✔
  client choosing clusterEndpoint http://localhost:60000/system +1s
  client Sending query +0ms
  client
  client       query ($projectId: ID!){
  client         viewer {
  client           project(id: $projectId) {
  client             ...RemoteProject
  client           }
  client         }
  client       }
  client
  client   fragment RemoteProject on Project {
  client     id
  client     name
  client     schema
  client     alias
  client     region
  client     isEjected
  client     projectDefinitionWithFileContent
  client   }
  client
  client        +0ms
  client { projectId: 'cj9xg08wk00040125983ve1g8' } +0ms
  deploy bundling +6s
  bundler /Users/alvincrespo/Desktop/derpity/.build/build.zip +2ms
Bundling functions...  bundler added files [ '.graphcoolrc',
  'graphcool.yml',
  'package.json',
  'src/',
  'types.graphql',
  '.build/build.zip',
  '.build/dist/',
  'src/hello.graphql' ] +11ms
  ts-builder starting compile [ '/Users/alvincrespo/Desktop/derpity/src/hello.js' ] +1ms
  ts-builder created program +305ms
  ts-builder emitted +245ms
  bundler converted files [ '/Users/alvincrespo/Desktop/derpity/src/hello.js' ] +2ms
  bundler createdFiles [ '/Users/alvincrespo/Desktop/derpity/.build/dist/src/hello.js' ] +0ms
  bundler adding /Users/alvincrespo/Desktop/derpity/.build/dist/src/ src/ +15ms
  bundler adding /Users/alvincrespo/Desktop/derpity/.build/dist/src/byline.js src/byline.js +0ms
  bundler adding /Users/alvincrespo/Desktop/derpity/.build/dist/src/hello-dev.js src/hello-dev.js +0ms
  bundler adding /Users/alvincrespo/Desktop/derpity/.build/dist/src/hello-lambda.js src/hello-lambda.js +0ms
  bundler adding /Users/alvincrespo/Desktop/derpity/.build/dist/src/hello.js src/hello.js +0ms
  bundler added build files [ 'src/',
  'src/byline.js',
  'src/hello-dev.js',
  'src/hello-lambda.js',
  'src/hello.js' ] +0ms
  client choosing clusterEndpoint http://localhost:60000/system +750ms
  client Sending query +0ms
  client
  client       mutation getUrl($projectId: String!) {
  client         getTemporaryDeployUrl(
  client           input: {
  client           projectId: $projectId
  client         }
  client       ) {
  client           url
  client         }
  client       }
  client      +0ms
  client { projectId: 'cj9xg08wk00040125983ve1g8' } +1ms
  bundler uploading to http://localhost:60050/functions/files/cj9xg08wk00040125983ve1g8/cj9xg0ajj000p0125eeu2ft1a +171ms
  bundler {
  bundler   "success" : true
  bundler } +320ms
  bundler bundled 1069 +0ms
 1.1s
  deploy bundled +1ms
Deploying locally...  client
  client
  client Sending service definition: +362ms
  client {
  client   "modules": [
  client     {
  client       "name": "",
  client       "content": "types: ./types.graphql\nfunctions:\n  hello:\n    handler:\n      code: src/hello.js\n    type: resolver\n    schema: src/hello.graphql\npermissions:\n  - operation: '*'\n",
  client       "files": {
  client         "./types.graphql": "# The following types define the data model of the example service\n# based on which the GraphQL API is generated\n\ntype User @model {\n  id: ID! @isUnique\n  name: String\n  dateOfBirth: DateTime\n\n  # Uncomment below - you can declare relations between models like this\n\n  # posts: [Post!]! @relation(name: \"UserPosts\")\n}\n\n\n# Uncomment the model below as well\n\n# type Post @model {\n#   id: ID! @isUnique\n#   title: String!\n#\n#   # Every relation also required a back-relation (to determine 1:1, 1:n or n:m)\n#   author: User! @relation(name: \"UserPosts\")\n# }\n",
  client         "src/hello.graphql": "type HelloPayload {\n  message: String!\n}\n\nextend type Query {\n  hello(name: String): HelloPayload\n}\n"
  client       },
  client       "externalFiles": {
  client         "src/hello.js": {
  client           "url": "http://localhost:60050/functions/files/cj9xg08wk00040125983ve1g8/cj9xg0ajj000p0125eeu2ft1a",
  client           "lambdaHandler": "src/hello-lambda.handle",
  client           "devHandler": "src/hello-dev.js"
  client         }
  client       }
  client     }
  client   ]
  client } +4ms
  client choosing clusterEndpoint http://localhost:60000/system +1ms
  client Sending query +0ms
  client       mutation($projectId: String!, $force: Boolean, $isDryRun: Boolean!, $config: String!) {
  client         push(input: {
  client           projectId: $projectId
  client           force: $force
  client           isDryRun: $isDryRun
  client           config: $config
  client           version: 1
  client         }) {
  client           migrationMessages {
  client             type
  client             action
  client             name
  client             description
  client             subDescriptions {
  client               type
  client               action
  client               name
  client               description
  client             }
  client           }
  client           errors {
  client             description
  client             type
  client             field
  client           }
  client           project {
  client             id
  client             name
  client             alias
  client             projectDefinitionWithFileContent
  client           }
  client         }
  client       }
  client      +0ms
  client { projectId: 'cj9xg08wk00040125983ve1g8',
  client   force: undefined,
  client   isDryRun: false,
  client   config: '{"modules":[{"name":"","content":"types: ./types.graphql\\nfunctions:\\n  hello:\\n    handler:\\n      code: src/hello.js\\n    type: resolver\\n    schema: src/hello.graphql\\npermissions:\\n  - operation: \'*\'\\n","files":{"./types.graphql":"# The following types define the data model of the example service\\n# based on which the GraphQL API is generated\\n\\ntype User @model {\\n  id: ID! @isUnique\\n  name: String\\n  dateOfBirth: DateTime\\n\\n  # Uncomment below - you can declare relations between models like this\\n\\n  # posts: [Post!]! @relation(name: \\"UserPosts\\")\\n}\\n\\n\\n# Uncomment the model below as well\\n\\n# type Post @model {\\n#   id: ID! @isUnique\\n#   title: String!\\n#\\n#   # Every relation also required a back-relation (to determine 1:1, 1:n or n:m)\\n#   author: User! @relation(name: \\"UserPosts\\")\\n# }\\n","src/hello.graphql":"type HelloPayload {\\n  message: String!\\n}\\n\\nextend type Query {\\n  hello(name: String): HelloPayload\\n}\\n"},"externalFiles":{"src/hello.js":{"url":"http://localhost:60050/functions/files/cj9xg08wk00040125983ve1g8/cj9xg0ajj000p0125eeu2ft1a","lambdaHandler":"src/hello-lambda.handle","devHandler":"src/hello-dev.js"}}}]}' } +0ms
  client undefined +1s
 1.3s

Success! Created the following service:

Types

  User
   + A new type with the name `User` is created.
   ├─ +  A new field with the name `name` and type `String` is created.
   └─ +  A new field with the name `dateOfBirth` and type `DateTime` is created.

Resolver Functions

  hello
   + A new resolver function with the name `hello` is created.

Permissions

  Wildcard Permission
   ? The wildcard permission for all operations is added.

Here are your GraphQL Endpoints:

  Simple API:        http://localhost:60000/simple/v1/cj9xg08wk00040125983ve1g8
  Relay API:         http://localhost:60000/relay/v1/cj9xg08wk00040125983ve1g8
  Subscriptions API: ws://localhost:60000/subscriptions/v1/cj9xg08wk00040125983ve1g8
```

**Comment by kbrandwijk** [comment]
@alvincrespo And you are still getting the same error running the hello function?

**Comment by alvincrespo** [comment]
@kbrandwijk yup, getting the same error as before.

**Comment by alvincrespo** [comment]
Do you think I should file a new issue?

**Comment by kbrandwijk** [comment]
Yes, because I can see in the deploy logs that it's not related to the Windows fix.

**Comment by alvincrespo** [comment]
Ok, opened up a new issue here: 

https://github.com/graphcool/framework/issues/1265

**Comment by radicand** [comment]
I ran into this issue with 0.9.2 on Mac, and discovered changing/removing the `console.log(e)` error logging to `console.log(e.message)` to work. There was something about attempting to log the whole error object that made it bomb out with this error. After that I had more errors, but they were issues in the [template] code that needed to be addressed -- now that I was able to get an actual error message to see what was wrong.

**Comment by Lurow** [comment]
Can confirm that logging objects instead of strings triggers the error, I had a fetch against an external auth-service and tried to log the answer. After removing the `console.log`, everything works as expected.

**Comment by marktani** [comment]
@radicand @Lurow please create a new bug report with reproduction information 🙂 
Specifically, the exact shape of the object would be very helpful to pin down the issue.

Thanks! 🙌

**Comment by radicand** [comment]
@marktani I'm not sure a new report makes sense - it's the same issue as described in the original posted description. What @Lurow and I were confirming is that that issue is due to lines like this: https://github.com/graphcool/templates/blob/master/auth/google/src/googleAuthentication.ts#L48 . Logging the Error object will result in `{"error":"Function returned invalid status code: 0. Raw body: empty.last"}.`

**Comment by kbrandwijk** [comment]
@radicand Function invocation being broken on Windows had to do with OS-specific path separators, and that was fixed and this issue closed. The fact that console.logging and object instead of a string also causes an error is a separate issue, so it makes sense to fix it based on a new issue, not based on an already closed one.

**Comment by radicand** [comment]
Fair point - I'll open and reference.

**Comment by maxdarque** [comment]
Not sure if it's connected but I got the following ` "error": "Function returned invalid status code: 0. Raw body: empty.last"` when I loaded code using `src: ./functions/code.ts`. 

I changed it to `src: functions/code.ts` and it worked.

**Comment by Boid-John** [comment]
I have this error, when running the subscription function via invoke-local I I receive no errors. When deployed the function worked correctly for about two days and now it only returns.
  "error": "Function returned invalid status code: 0. Raw body: empty.last"
I tried the suggestions above and the issue is not resolved. Debugging managed functions is very time consuming, I might just abandon managed functions and use hooks.
